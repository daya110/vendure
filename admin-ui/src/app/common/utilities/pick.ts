/**
 * Returns a new object which is a subject of the input, including only the specified properties.
 * Can be called with a single argument (array of props to pick), in which case it returns a partially
 * applied pick function.
 */
export function pick<T extends object, U extends T = any>(props: Array<keyof T>): (input: U) => T;
export function pick<T extends object, U extends T = any>(input: U, props: Array<keyof T>): T;
export function pick<T extends object, U extends T = any>(
    inputOrProps: U | Array<keyof T>,
    maybeProps?: Array<keyof T>,
): T | ((input: U) => T) {
    if (Array.isArray(inputOrProps)) {
        return (input: U) => _pick(input, inputOrProps);
    } else {
        return _pick(inputOrProps, maybeProps || []);
    }
}

function _pick<T extends object, U extends T = any>(input: U, props: Array<keyof T>): T {
    const output: any = {};
    for (const prop of props) {
        output[prop] = input[prop];
    }
    return output;
}
