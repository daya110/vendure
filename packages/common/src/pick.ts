/**
 * Returns a new object which is a subset of the input, including only the specified properties.
 * Can be called with a single argument (array of props to pick), in which case it returns a partially
 * applied pick function.
 */
export function pick<T extends string>(props: T[]): <U>(input: U) => Pick<U, Extract<keyof U, T>>;
export function pick<U, T extends keyof U>(input: U, props: T[]): { [K in T]: U[K] };
export function pick<U, T extends keyof U>(
    inputOrProps: U | T[],
    maybeProps?: T[],
): { [K in T]: U[K] } | ((input: U) => Pick<U, Extract<keyof U, T>>) {
    if (Array.isArray(inputOrProps)) {
        return (input: U) => _pick(input, inputOrProps);
    } else {
        return _pick(inputOrProps, maybeProps || []);
    }
}

function _pick<U, T extends keyof U>(input: U, props: T[]): { [K in T]: U[K] } {
    const output: any = {};
    for (const prop of props) {
        output[prop] = input[prop];
    }
    return output;
}
