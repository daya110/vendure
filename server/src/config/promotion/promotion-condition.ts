import { ConfigArg } from '../../../../shared/generated-types';

import { Order } from '../../entity/order/order.entity';
import { argsArrayToHash, ConfigArgs, ConfigArgValues } from '../common/config-args';

export type PromotionConditionArgType = 'int' | 'money' | 'string' | 'datetime' | 'boolean';
export type PromotionConditionArgs = ConfigArgs<PromotionConditionArgType>;
export type CheckPromotionConditionFn<T extends PromotionConditionArgs> = (
    order: Order,
    args: ConfigArgValues<T>,
) => boolean;

export class PromotionCondition<T extends PromotionConditionArgs = {}> {
    readonly code: string;
    readonly description: string;
    readonly args: PromotionConditionArgs;
    readonly priorityValue: number;
    private readonly checkFn: CheckPromotionConditionFn<T>;

    constructor(config: {
        args: T;
        check: CheckPromotionConditionFn<T>;
        code: string;
        description: string;
        priorityValue?: number;
    }) {
        this.code = config.code;
        this.description = config.description;
        this.args = config.args;
        this.checkFn = config.check;
        this.priorityValue = config.priorityValue || 0;
    }

    check(order: Order, args: ConfigArg[]) {
        return this.checkFn(order, argsArrayToHash<T>(args));
    }
}
