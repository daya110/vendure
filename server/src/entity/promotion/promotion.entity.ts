import { Adjustment, AdjustmentOperation, AdjustmentType } from 'shared/generated-types';
import { DeepPartial } from 'shared/shared-types';
import { Column, Entity, JoinTable, ManyToMany } from 'typeorm';

import { AdjustmentSource } from '../../common/types/adjustment-source';
import { VendureEntity } from '../base/base.entity';
import { Channel } from '../channel/channel.entity';

@Entity()
export class Promotion extends AdjustmentSource {
    type = AdjustmentType.PROMOTION;

    constructor(input?: DeepPartial<Promotion>) {
        super(input);
    }

    @Column() name: string;

    @Column() enabled: boolean;

    @ManyToMany(type => Channel)
    @JoinTable()
    channels: Channel[];

    @Column('simple-json') conditions: AdjustmentOperation[];

    @Column('simple-json') actions: AdjustmentOperation[];

    apply(): Adjustment {
        return {} as any;
    }

    test(): boolean {
        return false;
    }
}
