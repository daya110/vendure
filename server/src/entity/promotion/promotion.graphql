type Promotion implements Node {
    id: ID!
    createdAt: DateTime!
    updatedAt: DateTime!
    name: String!
    enabled: Boolean!
    conditions: [AdjustmentOperation!]!
    actions: [AdjustmentOperation!]!
}


type AdjustmentArg {
    name: String!
    type: String!
    value: String
}

type AdjustmentOperation {
    code: String!
    args: [AdjustmentArg!]!
    description: String!
}

input AdjustmentOperationInput {
    code: String!
    arguments: [String!]!
}

input CreatePromotionInput {
    name: String!
    enabled: Boolean!
    conditions: [AdjustmentOperationInput!]!
    actions: [AdjustmentOperationInput!]!
}

input UpdatePromotionInput {
    id: ID!
    name: String
    enabled: Boolean
    conditions: [AdjustmentOperationInput!]
    actions: [AdjustmentOperationInput!]
}
