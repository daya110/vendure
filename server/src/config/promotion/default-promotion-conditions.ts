import { Order } from '../../entity/order/order.entity';

import { PromotionCondition } from './promotion-condition';

export const minimumOrderAmount = new PromotionCondition({
    description: 'If order total is greater than { amount }',
    code: 'minimum_order_amount',
    args: {
        amount: 'money',
        taxInclusive: 'boolean',
    },
    check(order, args) {
        if (args.taxInclusive) {
            return order.subTotal >= args.amount;
        } else {
            return order.subTotalBeforeTax >= args.amount;
        }
    },
    priorityValue: 10,
});

export const dateRange = new PromotionCondition({
    code: 'date_range',
    description: 'If Order placed between { start } and { end }',
    args: { start: 'datetime', end: 'datetime' },
    check(order: Order, args) {
        const now = new Date();
        return args.start < now && now < args.end;
    },
});

export const atLeastNOfProduct = new PromotionCondition({
    code: 'at_least_n_of_product',
    description: 'Buy at least { minimum } of any product',
    args: { minimum: 'int' },
    check(order: Order, args) {
        return order.lines.reduce((result, item) => {
            return result || item.quantity >= args.minimum;
        }, false);
    },
});

export const atLeastNWithFacets = new PromotionCondition({
    code: 'at_least_n_with_facets',
    description: 'Buy at least { minimum } with the given facets',
    args: { minimum: 'int', facets: 'facetValueIds' },
    async check(order: Order, args, { hasFacetValues }) {
        let matches = 0;
        for (const line of order.lines) {
            if (await hasFacetValues(line, args.facets)) {
                matches += line.quantity;
            }
        }
        return args.minimum <= matches;
    },
});

export const defaultPromotionConditions = [
    minimumOrderAmount,
    dateRange,
    atLeastNOfProduct,
    atLeastNWithFacets,
];
