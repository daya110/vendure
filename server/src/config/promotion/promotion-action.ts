import { ConfigArg } from '../../../../shared/generated-types';
import {
    argsArrayToHash,
    ConfigArgs,
    ConfigArgValues,
    ConfigurableOperationDef,
} from '../../common/types/configurable-operation';
import { OrderItem } from '../../entity/order-item/order-item.entity';
import { OrderLine } from '../../entity/order-line/order-line.entity';
import { Order } from '../../entity/order/order.entity';

import { PromotionUtils } from './promotion-condition';

export type PromotionActionArgType = 'percentage' | 'money' | 'int' | 'facetValueIds';
export type PromotionActionArgs = ConfigArgs<PromotionActionArgType>;

export type ExecutePromotionItemActionFn<T extends PromotionActionArgs> = (
    orderItem: OrderItem,
    orderLine: OrderLine,
    args: ConfigArgValues<T>,
    utils: PromotionUtils,
) => number | Promise<number>;
export type ExecutePromotionOrderActionFn<T extends PromotionActionArgs> = (
    order: Order,
    args: ConfigArgValues<T>,
    utils: PromotionUtils,
) => number | Promise<number>;

export interface PromotionActionConfig<T extends PromotionActionArgs> {
    args: T;
    code: string;
    description: string;
    priorityValue?: number;
}
export interface PromotionItemActionConfig<T extends PromotionActionArgs> extends PromotionActionConfig<T> {
    execute: ExecutePromotionItemActionFn<T>;
}
export interface PromotionOrderActionConfig<T extends PromotionActionArgs> extends PromotionActionConfig<T> {
    execute: ExecutePromotionOrderActionFn<T>;
}

/**
 * @description
 * An abstract class which is extended by {@link PromotionItemAction} and {@link PromotionOrderAction}.
 *
 * @docsCategory promotions
 */
export abstract class PromotionAction<T extends PromotionActionArgs = {}>
    implements ConfigurableOperationDef {
    readonly code: string;
    readonly args: PromotionActionArgs;
    readonly description: string;
    readonly priorityValue: number;

    protected constructor(config: PromotionActionConfig<T>) {
        this.code = config.code;
        this.description = config.description;
        this.args = config.args;
        this.priorityValue = config.priorityValue || 0;
    }
}

/**
 * @description
 * Represents a PromotionAction which applies to individual {@link OrderItem}s.
 *
 * @example
 * ```ts
 * // Applies a percentage discount to each OrderItem
 * const itemPercentageDiscount = new PromotionItemAction({
 *     code: 'item_percentage_discount',
 *     args: { discount: 'percentage' },
 *     execute(orderItem, orderLine, args) {
 *         return -orderLine.unitPrice * (args.discount / 100);
 *     },
 *     description: 'Discount every item by { discount }%',
 * });
 * ```
 *
 * @docsCategory promotions
 */
export class PromotionItemAction<T extends PromotionActionArgs = {}> extends PromotionAction<T> {
    private readonly executeFn: ExecutePromotionItemActionFn<T>;
    constructor(config: PromotionItemActionConfig<T>) {
        super(config);
        this.executeFn = config.execute;
    }

    execute(orderItem: OrderItem, orderLine: OrderLine, args: ConfigArg[], utils: PromotionUtils) {
        return this.executeFn(orderItem, orderLine, argsArrayToHash(args), utils);
    }
}

/**
 * @description
 * Represents a PromotionAction which applies to the {@link Order} as a whole.
 *
 * @example
 * ```ts
 * // Applies a percentage discount to the entire Order
 * const orderPercentageDiscount = new PromotionOrderAction({
 *     code: 'order_percentage_discount',
 *     args: { discount: 'percentage' },
 *     execute(order, args) {
 *         return -order.subTotal * (args.discount / 100);
 *     },
 *     description: 'Discount order by { discount }%',
 * });
 * ```
 *
 * @docsCategory promotions
 */
export class PromotionOrderAction<T extends PromotionActionArgs = {}> extends PromotionAction<T> {
    private readonly executeFn: ExecutePromotionOrderActionFn<T>;
    constructor(config: PromotionOrderActionConfig<T>) {
        super(config);
        this.executeFn = config.execute;
    }

    execute(order: Order, args: ConfigArg[], utils: PromotionUtils) {
        return this.executeFn(order, argsArrayToHash(args), utils);
    }
}
