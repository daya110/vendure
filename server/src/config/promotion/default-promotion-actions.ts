import { PromotionAction, PromotionItemAction, PromotionOrderAction } from './promotion-action';

export const orderPercentageDiscount = new PromotionOrderAction({
    code: 'order_percentage_discount',
    args: { discount: 'percentage' },
    execute(order, args) {
        return -order.subTotal * (args.discount / 100);
    },
    description: 'Discount order by { discount }%',
});

export const itemPercentageDiscount = new PromotionItemAction({
    code: 'item_percentage_discount',
    args: { discount: 'percentage' },
    execute(orderItem, orderLine, args) {
        return -orderLine.unitPrice * (args.discount / 100);
    },
    description: 'Discount every item by { discount }%',
});

export const buy1Get1Free = new PromotionItemAction({
    code: 'buy_1_get_1_free',
    args: {},
    execute(orderItem, orderLine, args) {
        if (orderLine.quantity >= 2) {
            const lineIndex = orderLine.items.indexOf(orderItem) + 1;
            if (lineIndex % 2 === 0) {
                return -orderLine.unitPrice;
            }
        }
        return 0;
    },
    description: 'Discount every item by { discount }%',
});

export const defaultPromotionActions = [orderPercentageDiscount, itemPercentageDiscount, buy1Get1Free];
