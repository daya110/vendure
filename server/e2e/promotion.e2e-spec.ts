import {
    CreatePromotion,
    GetAdjustmentOperations,
    GetPromotion,
    GetPromotionList,
    Promotion,
    UpdatePromotion,
} from 'shared/generated-types';
import { pick } from 'shared/pick';

import {
    CREATE_PROMOTION,
    GET_ADJUSTMENT_OPERATIONS,
    GET_PROMOTION,
    GET_PROMOTION_LIST,
    UPDATE_PROMOTION,
} from '../../admin-ui/src/app/data/definitions/promotion-definitions';
import { PromotionAction } from '../src/config/promotion/promotion-action';
import { PromotionCondition } from '../src/config/promotion/promotion-condition';

import { TestClient } from './test-client';
import { TestServer } from './test-server';

// tslint:disable:no-non-null-assertion

describe('Promotion resolver', () => {
    const client = new TestClient();
    const server = new TestServer();

    const promoCondition = generateTestCondition('promo_condition');
    const promoCondition2 = generateTestCondition('promo_condition2');

    const promoAction = generateTestAction('promo_action');

    const snapshotProps = ['name', 'actions', 'conditions', 'enabled'] as Array<
        'name' | 'actions' | 'conditions' | 'enabled'
    >;
    let promotion: Promotion.Fragment;

    beforeAll(async () => {
        const token = await server.init(
            {
                productCount: 1,
                customerCount: 1,
            },
            {
                promotionConditions: [promoCondition, promoCondition2],
                promotionActions: [promoAction],
            },
        );
        await client.init();
    }, 60000);

    afterAll(async () => {
        await server.destroy();
    });

    it('createPromotion promotion', async () => {
        const result = await client.query<CreatePromotion.Mutation, CreatePromotion.Variables>(
            CREATE_PROMOTION,
            {
                input: {
                    name: 'test promotion',
                    enabled: true,
                    conditions: [
                        {
                            code: promoCondition.code,
                            arguments: [{ name: 'arg', value: '500' }],
                        },
                    ],
                    actions: [
                        {
                            code: promoAction.code,
                            arguments: [{ name: 'percentage', value: '50' }],
                        },
                    ],
                },
            },
        );
        promotion = result.createPromotion;
        expect(pick(promotion, snapshotProps)).toMatchSnapshot();
    });

    it('updatePromotion', async () => {
        const result = await client.query<UpdatePromotion.Mutation, UpdatePromotion.Variables>(
            UPDATE_PROMOTION,
            {
                input: {
                    id: promotion.id,
                    conditions: [
                        {
                            code: promoCondition.code,
                            arguments: [{ name: 'arg', value: '90' }],
                        },
                        {
                            code: promoCondition2.code,
                            arguments: [{ name: 'arg', value: '10' }],
                        },
                    ],
                },
            },
        );
        expect(pick(result.updatePromotion, snapshotProps)).toMatchSnapshot();
    });

    it('promotion', async () => {
        const result = await client.query<GetPromotion.Query, GetPromotion.Variables>(GET_PROMOTION, {
            id: promotion.id,
        });

        expect(result.promotion!.name).toBe(promotion.name);
    });

    it('promotions', async () => {
        const result = await client.query<GetPromotionList.Query, GetPromotionList.Variables>(
            GET_PROMOTION_LIST,
            {},
        );

        expect(result.promotions.totalItems).toBe(1);
        expect(result.promotions.items[0].name).toBe('test promotion');
    });

    it('adjustmentOperations', async () => {
        const result = await client.query<GetAdjustmentOperations.Query, GetAdjustmentOperations.Variables>(
            GET_ADJUSTMENT_OPERATIONS,
        );

        expect(result.adjustmentOperations).toMatchSnapshot();
    });
});

function generateTestCondition(code: string): PromotionCondition<any> {
    return new PromotionCondition({
        code,
        description: `description for ${code}`,
        args: { arg: 'money' },
        check: (order, args) => true,
    });
}

function generateTestAction(code: string): PromotionAction<any> {
    return new PromotionAction({
        code,
        description: `description for ${code}`,
        args: { percentage: 'percentage' },
        execute: (order, args) => {
            return 42;
        },
    });
}
