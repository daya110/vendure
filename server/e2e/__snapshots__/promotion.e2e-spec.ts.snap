// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Promotion resolver adjustmentOperations 1`] = `
Object {
  "actions": Array [
    Object {
      "args": Array [
        Object {
          "name": "percentage",
          "type": "PERCENTAGE",
          "value": null,
        },
      ],
      "code": "promo_action",
      "description": "description for promo_action",
    },
  ],
  "conditions": Array [
    Object {
      "args": Array [
        Object {
          "name": "arg",
          "type": "MONEY",
          "value": null,
        },
      ],
      "code": "promo_condition",
      "description": "description for promo_condition",
    },
    Object {
      "args": Array [
        Object {
          "name": "arg",
          "type": "MONEY",
          "value": null,
        },
      ],
      "code": "promo_condition2",
      "description": "description for promo_condition2",
    },
  ],
}
`;

exports[`Promotion resolver createPromotion 1`] = `
Object {
  "actions": Array [
    Object {
      "args": Array [
        Object {
          "name": "percentage",
          "type": "PERCENTAGE",
          "value": "50",
        },
      ],
      "code": "promo_action",
      "description": "description for promo_action",
    },
  ],
  "conditions": Array [
    Object {
      "args": Array [
        Object {
          "name": "arg",
          "type": "MONEY",
          "value": "500",
        },
      ],
      "code": "promo_condition",
      "description": "description for promo_condition",
    },
  ],
  "enabled": true,
  "name": "test promotion",
}
`;

exports[`Promotion resolver updatePromotion 1`] = `
Object {
  "actions": Array [
    Object {
      "args": Array [
        Object {
          "name": "percentage",
          "type": "PERCENTAGE",
          "value": "50",
        },
      ],
      "code": "promo_action",
      "description": "description for promo_action",
    },
  ],
  "conditions": Array [
    Object {
      "args": Array [
        Object {
          "name": "arg",
          "type": "MONEY",
          "value": "90",
        },
      ],
      "code": "promo_condition",
      "description": "description for promo_condition",
    },
    Object {
      "args": Array [
        Object {
          "name": "arg",
          "type": "MONEY",
          "value": "10",
        },
      ],
      "code": "promo_condition2",
      "description": "description for promo_condition2",
    },
  ],
  "enabled": true,
  "name": "test promotion",
}
`;
